Adding DeepSort object tracking utility

rohis06 opened the pull request:
## What changes are proposed in this pull request?

Adds DeepSort object tracking utility to FiftyOne 🎊 

It can be used as follows:

```python
!pip3 install deep-sort-realtime

import fiftyone as fo
import fiftyone.zoo as foz
from fiftyone import ViewField as F
from fiftyone.utils.tracking import DeepSort
from ultralytics import YOLO

model = YOLO('yolov8n.pt')

dataset = foz.load_zoo_dataset("quickstart-video", dataset_name="qv", max_samples=3)

dataset.ensure_frames()

dataset.apply_model(model, label_field="frames.detections")

DeepSort.track(dataset=dataset, in_field="frames.detections", max_age=15, progress=True)

session = fo.launch_app(dataset)

trajectories = (
    dataset
    .filter_labels("frames.ds_tracks", F("label") == "car")
    .to_trajectories("frames.ds_tracks")
)

session = fo.launch_app(view=trajectories)
```

## How is this patch tested? If it is not, please explain why.

It has been thoroughly tested using the above code snippet.

## Release Notes

### Is this a user-facing change that should be mentioned in the release notes?

<!--
Please fill in relevant options below with an "x", or by clicking the checkboxes
after submitting this pull request. Example:
-   [x] Selected option
-->

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

Adds DeepSort object tracking utility to FiftyOne.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

jacobmarks commented:
This is awesome @rohis06 !!! I love the simplicity of your implementation, and this is a huge upgrade for FiftyOne 💪. The way you laid the structure out to accommodate others trackers in the future is perfect. The implementation works for me!!! 

Just one very minor question/note: the tracks have index numbers that are pretty high — it looks like a lot of the tracks got cut and the index is never reset. This means that the cars have labels like "car 14", "car 18", ... as opposed to "car 0", "car 1", ... etc. Is this the convention, if not, it might be nice to parse through the index numbers and reassign numbers starting from 0. What do you think?

Also, I LOVE the illustration of trajectories. Thank you for making this by far more useful than ever before!!

rohis06 commented:
Thanks for reviewing and providing your feedback, @jacobmarks! 😄
 I am glad that it met the expectations 😊 
 
 > Just one very minor question/note: the tracks have index numbers that are pretty high — it looks like a lot of the tracks got cut and the index is never reset. This means that the cars have labels like "car 14", "car 18", ... as opposed to "car 0", "car 1", ... etc. Is this the convention, if not, it might be nice to parse through the index numbers and reassign numbers starting from 0. What do you think?

That's actually correct since the DeepSort algorithm outputs the global tracking IDs. So, "car 14" doesn't essentially translate to the 14th car. Instead, it's the 14th object detected and tracked across all the frames of the input video. Even I felt it to be a little misleading. So, do you think it would be better to update the code so that the `out_field` only contains the bounding_box and index (tracking_id) of the detection and not include the labels and confidence?

I can update the lines 112-119 to:

```python
                    tracked_detections.append(
                        fo.Detection(
                            bounding_box=[rel_x, rel_y, rel_w, rel_h],
                            index=track.track_id,
                        )
                    )
```

Let me know what you think about this approach.

jacobmarks commented:
> Thanks for reviewing and providing your feedback, @jacobmarks! 😄 I am glad that it met the expectations 😊
> 
> > Just one very minor question/note: the tracks have index numbers that are pretty high — it looks like a lot of the tracks got cut and the index is never reset. This means that the cars have labels like "car 14", "car 18", ... as opposed to "car 0", "car 1", ... etc. Is this the convention, if not, it might be nice to parse through the index numbers and reassign numbers starting from 0. What do you think?
> 
> That's actually correct since the DeepSort algorithm outputs the global tracking IDs. So, "car 14" doesn't essentially translate to the 14th car. Instead, it's the 14th object detected and tracked across all the frames of the input video. Even I felt it to be a little misleading. So, do you think it would be better to update the code so that the `out_field` only contains the bounding_box and index (tracking_id) of the detection and not include the labels and confidence?
> 
> I can update the lines 112-119 to:
> 
> ```python
>                     tracked_detections.append(
>                         fo.Detection(
>                             bounding_box=[rel_x, rel_y, rel_w, rel_h],
>                             index=track.track_id,
>                         )
>                     )
> ```
> 
> Let me know what you think about this approach.

Hmm I see. I think the best thing to do would be to add an argument to the tracker like `keep_confidence=False` which allows the user to decide whether they want this. What do you think?

jacobmarks commented:
@rohis06 I'm also seeing some small issues with bounding boxes/tracks.
<img width="1724" alt="Screenshot 2024-04-20 at 11 13 07 AM" src="https://github.com/voxel51/fiftyone/assets/12500356/56fa543b-597f-4b91-a75a-d344ff61012d">

1. It looks like the tracks don't start until frame 3 for each sample. 
2. Looks like some of the tracks from sample 1 and propagating through to sample 2. Screenshot attached. This might be user error though!

rohis06 commented:
> > Thanks for reviewing and providing your feedback, @jacobmarks! 😄 I am glad that it met the expectations 😊
> > > Just one very minor question/note: the tracks have index numbers that are pretty high — it looks like a lot of the tracks got cut and the index is never reset. This means that the cars have labels like "car 14", "car 18", ... as opposed to "car 0", "car 1", ... etc. Is this the convention, if not, it might be nice to parse through the index numbers and reassign numbers starting from 0. What do you think?
> > 
> > 
> > That's actually correct since the DeepSort algorithm outputs the global tracking IDs. So, "car 14" doesn't essentially translate to the 14th car. Instead, it's the 14th object detected and tracked across all the frames of the input video. Even I felt it to be a little misleading. So, do you think it would be better to update the code so that the `out_field` only contains the bounding_box and index (tracking_id) of the detection and not include the labels and confidence?
> > I can update the lines 112-119 to:
> > ```python
> >                     tracked_detections.append(
> >                         fo.Detection(
> >                             bounding_box=[rel_x, rel_y, rel_w, rel_h],
> >                             index=track.track_id,
> >                         )
> >                     )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Let me know what you think about this approach.
> 
> Hmm I see. I think the best thing to do would be to add an argument to the tracker like `keep_confidence=False` which allows the user to decide whether they want this. What do you think?

Yeah, this sounds like a great approach. I will update the code to include the `keep_confidence=False` argument.

rohis06 commented:
> @rohis06 I'm also seeing some small issues with bounding boxes/tracks. <img alt="Screenshot 2024-04-20 at 11 13 07 AM" width="1724" src="https://private-user-images.githubusercontent.com/12500356/324186443-56fa543b-597f-4b91-a75a-d344ff61012d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM3MzQxMTEsIm5iZiI6MTcxMzczMzgxMSwicGF0aCI6Ii8xMjUwMDM1Ni8zMjQxODY0NDMtNTZmYTU0M2ItNTk3Zi00YjkxLWE3NWEtZDM0NGZmNjEwMTJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIxVDIxMTAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjZGZiODIwOGYwZGJhZTc5NGI5ZDI4MzJmOTM4ZmVjNjc5ZGM5M2IyMDMxNDE4MjJiOTIwODJkMzg1MjhmMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5vOUiPupPQ7CQb_M-mtcm9teWYgnt8-__ruu7sK9mGY">
> 
> 1. It looks like the tracks don't start until frame 3 for each sample.
> 2. Looks like some of the tracks from sample 1 and propagating through to sample 2. Screenshot attached. This might be user error though!

Hmm, I see. Could you please share the code snippet that resulted in this issue? I will try to reproduce it at my end and debug it further.

jacobmarks commented:
> > @rohis06 I'm also seeing some small issues with bounding boxes/tracks. <img alt="Screenshot 2024-04-20 at 11 13 07 AM" width="1724" src="https://private-user-images.githubusercontent.com/12500356/324186443-56fa543b-597f-4b91-a75a-d344ff61012d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM3MzQxMTEsIm5iZiI6MTcxMzczMzgxMSwicGF0aCI6Ii8xMjUwMDM1Ni8zMjQxODY0NDMtNTZmYTU0M2ItNTk3Zi00YjkxLWE3NWEtZDM0NGZmNjEwMTJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIxVDIxMTAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjZGZiODIwOGYwZGJhZTc5NGI5ZDI4MzJmOTM4ZmVjNjc5ZGM5M2IyMDMxNDE4MjJiOTIwODJkMzg1MjhmMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5vOUiPupPQ7CQb_M-mtcm9teWYgnt8-__ruu7sK9mGY">
> > 
> > 1. It looks like the tracks don't start until frame 3 for each sample.
> > 2. Looks like some of the tracks from sample 1 and propagating through to sample 2. Screenshot attached. This might be user error though!
> 
> Hmm, I see. Could you please share the code snippet that resulted in this issue? I will try to reproduce it at my end and debug it further.

Thanks for looking into this!

I just used this code snippet:

```py
import fiftyone as fo
import fiftyone.zoo as foz
import fiftyone.brain as fob
from fiftyone import ViewField as F
from fiftyone.utils.tracking import DeepSort

dataset = foz.load_zoo_dataset("quickstart-video", max_samples=3)
model = foz.load_zoo_model("yolov8n-coco-torch")

dataset.ensure_frames()
dataset.apply_model(model, label_field="frames.yolo_detection")

DeepSort.track(dataset=dataset, in_field="frames.yolo_detection", max_age=15, progress=True)
```

rohis06 commented:
> @rohis06 I'm also seeing some small issues with bounding boxes/tracks. <img alt="Screenshot 2024-04-20 at 11 13 07 AM" width="1724" src="https://private-user-images.githubusercontent.com/12500356/324186443-56fa543b-597f-4b91-a75a-d344ff61012d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM4NTEzMzIsIm5iZiI6MTcxMzg1MTAzMiwicGF0aCI6Ii8xMjUwMDM1Ni8zMjQxODY0NDMtNTZmYTU0M2ItNTk3Zi00YjkxLWE3NWEtZDM0NGZmNjEwMTJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIzVDA1NDM1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2YzBlYWQzMTBhMTEyZWFjYWRkYmYwNGQ3ZWI5MzAyNDNmYmYwOWRlMmJjZmQxMzIzY2VkNWUwMjBiYThlZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vCb2l658008MLPERGx5HuwPpirPYTzavU6miA5q3WAI">
> 
> 1. It looks like the tracks don't start until frame 3 for each sample.

So, according to the official implementation of the deep sort algorithm, tracks have [three states](https://github.com/nwojke/deep_sort/blob/280b8bdb255f223813ff4a8679f3e1321b08cdfc/deep_sort/track.py#L160): Tentative, Confirmed, and Deleted. If we look at line 102 in our code, we can see that we ignore the tracks that aren't "confirmed" yet, which means not enough evidence has been collected for these tracks yet. In this specific dataset that we are testing, it looks like only when the 3rd frame is seen the tracks change their state from "tentative" to "confirmed". However, it doesn't always have to be the 3rd frame.

> 2. Looks like some of the tracks from sample 1 and propagating through to sample 2. Screenshot attached. This might be user error though!

Thanks for pointing this out, @jacobmarks. I have addressed this issue.

jacobmarks commented:
Awesome job @rohis06 , just ran the latest and it works swimmingly :) Massive kudos 🚀

jacobmarks reviewed (approved):
👍  LGTM!!!

rohis06 commented:
> Awesome job @rohis06 , just ran the latest and it works swimmingly :) Massive kudos 🚀

Thanks so much! Glad to hear it's working well! 🚀
